ethereum/contracts/BasicInboundChannel.sol

```diff
@@ -24,39 +27,34 @@ contract BasicInboundChannel {
 
     event MessageDispatched(uint64 id, bool result);
 
-    constructor(ParachainClient client) {
+    constructor(uint8 _sourceChannelID, ParachainClient _parachainClient) {
         nonce = 0;
-        parachainClient = client;
+        sourceChannelID = _sourceChannelID;
+        parachainClient = _parachainClient;
     }
 
-    function submit(MessageBundle calldata bundle, ParachainClient.Proof calldata proof) external {
+    function submit(MessageBundle calldata bundle,  bytes calldata proof) external {
         bytes32 commitment = keccak256(abi.encode(bundle));
 
         require(parachainClient.verifyCommitment(commitment, proof), "Invalid proof");
-
-        // Require there is enough gas to play all messages
+        require(bundle.sourceChannelID == sourceChannelID, "Invalid source channel");
+        require(bundle.nonce == nonce + 1, "Invalid nonce");
         require(
             gasleft() >= (bundle.messages.length * MAX_GAS_PER_MESSAGE) + GAS_BUFFER,
             "insufficient gas for delivery of all messages"
         );
 
-        processMessages(bundle);
+        dispatch(bundle);
+        nonce++;
```

Comment:
Is it not better practice to increment the nonce before calling other methods just in case of re-entrancy? `dispatch` calls `message.target` which could potentially be re-entrant and allow code malicious code to be executed before the nonce is incremented?

> ensure that all logic that changes state variables happen before ether is sent out of the contract